src/hotspot/share/cds/aotLinkedClassBulkLoader.cpp

```diff
@@ -0,0 +1,210 @@
+  if (classes == nullptr) {
+    return;
+  }
+
+  ClassLoaderData* loader_data = ClassLoaderData::class_loader_data(loader());
+
+  for (int i = 0; i < classes->length(); i++) {
+    InstanceKlass* ik = classes->at(i);
+    if (log_is_enabled(Info, cds, aot, load)) {
+      ResourceMark rm(THREAD);
+      log_info(cds, aot, load)("%s %s%s%s", category, ik->external_name(),
+                               ik->is_loaded() ? " (already loaded)" : "",
+                               ik->is_hidden() ? " (hidden)" : "");
+    }
+
+    if (!ik->is_loaded()) {
+      if (ik->is_hidden()) {
+        // TODO: AOTClassLinking is not implemented for hidden class until JDK-8293336
+        ShouldNotReachHere();
+      } else {
+        InstanceKlass* actual;
+        if (loader_data == ClassLoaderData::the_null_class_loader_data()) {
+          actual = SystemDictionary::load_instance_class(ik->name(), loader, CHECK);
+        } else {
+          actual = SystemDictionaryShared::find_or_load_shared_class(ik->name(), loader, CHECK);
+        }
+
+        if (actual != ik) {
+          ResourceMark rm(THREAD);
+          log_error(cds)("Unable to resolve %s class from CDS archive: %s", category, ik->external_name());
+          log_error(cds)("Expected: " INTPTR_FORMAT ", actual: " INTPTR_FORMAT, p2i(ik), p2i(actual));
+          log_error(cds)("JVMTI class retransformation is not supported when archive was generated with -XX:+AOTClassLinking.");
+          MetaspaceShared::unrecoverable_loading_error();
+        }
+        assert(actual->is_loaded(), "must be");
+      }
+    }
+  }
+}
+
+// Initiate loading of the <classes> in the <initiating_loader>. The <classes> should have already been loaded
+// by a parent loader of the <initiating_loader>. This is necessary for handling pre-resolved CP entries.
+//
+// For example, we initiate the loading of java/lang/String in the AppClassLoader. This will allow
+// any App classes to have a pre-resolved ConstantPool entry that references java/lang/String.
+//
+// TODO: we can limit the number of initiated classes to only those that are actually referenced by
+// AOT-linked classes loaded by <initiating_loader>.
+void AOTLinkedClassBulkLoader::initiate_loading(JavaThread* current, const char* category,
+                                                Handle initiating_loader, Array<InstanceKlass*>* classes) {
+  if (classes == nullptr) {
+    return;
+  }
+
+  assert(initiating_loader() == SystemDictionary::java_platform_loader() ||
+         initiating_loader() == SystemDictionary::java_system_loader(), "must be");
+  ClassLoaderData* loader_data = ClassLoaderData::class_loader_data(initiating_loader());
+  MonitorLocker mu1(SystemDictionary_lock);
+
+  for (int i = 0; i < classes->length(); i++) {
+    InstanceKlass* ik = classes->at(i);
+    assert(ik->is_loaded(), "must have already been loaded by a parent loader");
+    assert(ik->class_loader() != initiating_loader(), "must be a parent loader");
```

Comment:
Can we also add an assert that ik->class_loader() must be either boot or platform loader.

Reply:
Done.